txnsync/txngroups.go

```diff
@@ -17,23 +17,417 @@
 }
 
-func encodeTransactionGroups(inTxnGroups []transactions.SignedTxGroup) []byte {
+type encodedSignedTxns struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Sig             []byte                  `codec:"sig,allocbound=maxSignatureBytes"`
+	BitmaskSig      bitmask                 `codec:"sigbm"`
+	Msig            []crypto.MultisigSig    `codec:"msig,allocbound=maxEncodedTransactionGroup"`
+	BitmaskMsig     bitmask                 `codec:"msigbm"`
+	Lsig            []transactions.LogicSig `codec:"lsig,allocbound=maxEncodedTransactionGroup"`
+	BitmaskLsig     bitmask                 `codec:"lsigbm"`
+	AuthAddr        []byte                  `codec:"sgnr,allocbound=maxAddressBytes"`
+	BitmaskAuthAddr bitmask                 `codec:"sgnrbm"`
+
+	encodedTxns
+}
+
+type encodedTxns struct {
+	_struct       struct{} `codec:",omitempty,omitemptyarray"`
+	TxType        []byte   `codec:"type,allocbound=maxEncodedTransactionGroup"`
+	BitmaskTxType bitmask  `codec:"typebm"`
+
+	encodedTxnHeaders
+	encodedKeyregTxnFields
+	encodedPaymentTxnFields
+	encodedAssetConfigTxnFields
+	encodedAssetTransferTxnFields
+	encodedAssetFreezeTxnFields
+	encodedApplicationCallTxnFields
+	encodedCompactCertTxnFields
+}
+
+type encodedTxnHeaders struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Sender             []byte              `codec:"snd,allocbound=maxAddressBytes"`
+	BitmaskSender      bitmask             `codec:"sndbm"`
+	Fee                []basics.MicroAlgos `codec:"fee,allocbound=maxEncodedTransactionGroup"`
+	BitmaskFee         bitmask             `codec:"feebm"`
+	FirstValid         []basics.Round      `codec:"fv,allocbound=maxEncodedTransactionGroup"`
+	BitmaskFirstValid  bitmask             `codec:"fvbm"`
+	LastValid          []basics.Round      `codec:"lv,allocbound=maxEncodedTransactionGroup"`
+	BitmaskLastValid   bitmask             `codec:"lvbm"`
+	Note               [][]byte            `codec:"note,allocbound=maxEncodedTransactionGroup"` // TODO whats the correct allocbound?
+	BitmaskNote        bitmask             `codec:"notebm"`
+	GenesisID          string              `codec:"gen"`
+	BitmaskGenesisID   bitmask             `codec:"genbm"`
+	GenesisHash        crypto.Digest       `codec:"gh"`
+	BitmaskGenesisHash bitmask             `codec:"ghbm"`
+
+	BitmaskGroup bitmask `codec:"grpbm"`
+
+	Lease        []byte  `codec:"lx,allocbound=maxAddressBytes"`
+	BitmaskLease bitmask `codec:"lxbm"`
+
+	RekeyTo        []byte  `codec:"rekey,allocbound=maxAddressBytes"`
+	BitmaskRekeyTo bitmask `codec:"rekeybm"`
+}
+
+type encodedKeyregTxnFields struct {
+	_struct                 struct{}       `codec:",omitempty,omitemptyarray"`
+	VotePK                  []byte         `codec:"votekey,allocbound=maxAddressBytes"`
+	BitmaskVotePK           bitmask        `codec:"votekeybm"`
+	SelectionPK             []byte         `codec:"selkey,allocbound=maxAddressBytes"`
+	BitmaskSelectionPK      bitmask        `codec:"selkeybm"`
```

Comment:
The `SelectionPK`, `VotePK` and `KeyDilution` would always share the same bitmask.